lambda/transl_list_comprehension.ml

```diff
@@ -0,0 +1,302 @@
+   building the final list.
+
+   For example, consider the following list comprehension:
+   {[
+     [x+y for x = 1 to 3 when x <> 2 for y in [10*x; 100*x]]
+     (* = [11; 101; 33; 303] *)
+   ]}
+   This translates to the (Lambda equivalent of) the following:
+   {[
+     (* Convert the result to a normal list *)
+     CamlinternalComprehension.rev_list_to_list (
+       (* for x = 1 to 3 *)
+       let start = 1 in
+       let stop  = 3 in
+       CamlinternalComprehension.rev_dlist_concat_iterate_up
+         start stop
+         (fun x acc_x -> local_
+           (* when x <> 2 *)
+           if x <> 2
+           then
+             (* for y in [10*x; 100*x] *)
+             let iter_list = [10*x; 100*x] in
+             CamlinternalComprehension.rev_dlist_concat_map
+               iter_list
+               (fun y acc_y -> local_
+                 (* The body: x+y *)
+                 Snoc { init = acc_y; last = x*y })
+               acc_x
+           else
+             acc_x)
+         Nil)
+        ]}
+
+   See [CamlinternalComprehension] the types and functions we desugar to, along
+   with some more documentation. *)
+
+(** An implementation note: Many of the functions in this file need to translate
+    expressions from Typedtree to Lambda; to avoid strange dependency ordering,
+    we parameterize those functions by [Translcore.transl_exp], and pass it in
+    as a labeled argument, along with the necessary [scopes] labeled argument
+    that it requires. *)
+
+(* CR aspectorzabusky: I couldn't get this to build if these were run as soon as
+   this file was processed *)
+(** The functions that are required to build the results of list comprehensions;
+    see the documentation for [CamlinternalComprehension] for more details. *)
+let ( rev_list_to_list
+    , rev_dlist_concat_map
+    , rev_dlist_concat_iterate_up
+    , rev_dlist_concat_iterate_down )
+  =
+  let transl name =
+    lazy (Lambda.transl_prim "CamlinternalComprehension" name)
+  in
+  ( transl "rev_list_to_list"
+  , transl "rev_dlist_concat_map"
+  , transl "rev_dlist_concat_iterate_up"
+  , transl "rev_dlist_concat_iterate_down" )
+;;
+
+(** The [local_] form of the [CamlinternalComprehension.Snoc] constructor, for
+    building the intermediate restults of list comprehensions; see the
+    documentation for [CamlinternalComprehension.rev_list] for more details. *)
+let rev_list_snoc_local ~loc ~init ~last =
+  Lprim(Pmakeblock(0, Immutable, None, Lambda.alloc_local), [init; last], loc)
+
+(** The [CamlinternalComprehension.Nil] constructor, for building the
+    intermediate restults of list comprehensions; see the documentation for
+    [CamlinternalComprehension.rev_list] for more details. *)
+let rev_list_nil = int 0
+
+(** The information needed to translate a single iterator from a
+    [for ... and ...] clause (i.e., [x = e1 (down)to e2] or [for pat in xs]). *)
+type translated_iterator =
+  { builder : lambda Lazy.t
```

Comment:
Because they're bound to the functions above that are tagged "I couldn't get this to build if these were run as soon as this file was processed".  Once we figure out how to fix that I'll drop this.